source/extensions/access_loggers/fluentd/fluentd_access_log_impl.cc

```diff
@@ -0,0 +1,163 @@
+#include "source/extensions/access_loggers/fluentd/fluentd_access_log_impl.h"
+
+#include "source/common/buffer/buffer_impl.h"
+
+#include "msgpack.hpp"
+
+namespace Envoy {
+namespace Extensions {
+namespace AccessLoggers {
+namespace Fluentd {
+
+using MessagePackBuffer = msgpack::sbuffer;
+using MessagePackPacker = msgpack::packer<msgpack::sbuffer>;
+
+FluentdAccessLoggerImpl::FluentdAccessLoggerImpl(Tcp::AsyncTcpClientPtr client,
+                                                 Event::Dispatcher& dispatcher,
+                                                 const FluentdAccessLogConfig& config,
+                                                 Stats::Scope& parent_scope)
+    : tag_(config.tag()), id_(dispatcher.name()),
+      stats_scope_(parent_scope.createScope(config.stat_prefix())),
+      fluentd_stats_(
+          {ACCESS_LOG_FLUENTD_STATS(POOL_COUNTER(*stats_scope_), POOL_GAUGE(*stats_scope_))}),
+      client_(std::move(client)),
+      buffer_flush_interval_msec_(PROTOBUF_GET_MS_OR_DEFAULT(config, buffer_flush_interval, 1000)),
+      max_buffer_size_bytes_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config, buffer_size_bytes, 16384)),
+      flush_timer_(dispatcher.createTimer([this]() {
+        flush();
+        flush_timer_->enableTimer(buffer_flush_interval_msec_);
+      })) {
+  client_->setAsyncTcpClientCallbacks(*this);
+  flush_timer_->enableTimer(buffer_flush_interval_msec_);
+}
+
+void FluentdAccessLoggerImpl::onEvent(Network::ConnectionEvent event) {
+  connecting_ = false;
+
+  if (event == Network::ConnectionEvent::Connected) {
+    flush();
```

Comment:
or update docs on when flushing can occur

Reply:
We won't get here unless ``flush()`` has been previously called. On the first time ``flush()`` is called, and it has entries to send, it will first connect to the upstream and ``return;``, once the upstream is connected, ``onEvent(Connected)`` is called, and calls again to ``flush()`` to complete the first logging cycle.

Regarding docs - please see the comments of ``buffer_size_bytes`` and ``buffer_flush_interval``:
- ``// Interval for flushing access logs to the gRPC stream. Logger will flush requests every time
  // this interval is elapsed, or when batch size limit is hit, whichever comes first. Defaults to
  // 1 second.``
- ``// Soft size limit in bytes for access log entries buffer. The logger will buffer requests until
  // this limit it hit, or every time flush interval is elapsed, whichever comes first. When the buffer
  // limit is hit, the logger will immediately flush the buffer contents. Setting it to zero effectively
  // disables the batching. Defaults to 16384.``

Pls lmk if any additional comment is required